### PR TITLE
CPP: Add syntax examples to QLDoc in Function.qll

### PR DESCRIPTION
**Pull request by geoffw0**
For https://jira.semmle.com/browse/CPP-407.

Lots of classes in this file - took a while.

Some extra tests as well.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/Function.qll [hunk 1]
```diff
@@ -943,9 +1054,18 @@ class CopyConstructor extends Constructor {
 }
 
 /**
- * A C++ move constructor, such as `T::T(T&&)` [N4140 12.8].
+ * A C++ move constructor [N4140 12.8]. For example the function `MyClass` in
+ * the following code is a `MoveConstructor`:
+ * ```
+ * class MyClass {
+ * public:
+ *   MyClass(const MyClass &&from) {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/Function.qll]: There should not be `const` here. See https://en.cppreference.com/w/cpp/language/move_constructor.

File: cpp/ql/src/semmle/code/cpp/Function.qll [hunk 2]
```diff
@@ -1079,8 +1231,14 @@ class CopyAssignmentOperator extends Operator {
...
 
 /**
- * A C++ move assignment operator, such as `T& T::operator=(T&&)` [N4140
- * 12.8].
+ * A C++ move assignment operator [N4140 12.8]. For example the function
+ * `operator=` in the following code is a `MoveAssignmentOperator`:
+ * ```
+ * class MyClass {
+ * public:
+ *   MyClass &operator=(const MyClass &&other);
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/Function.qll]: Again, there should not be `const`.

File: cpp/ql/src/semmle/code/cpp/Function.qll [hunk 3]
```diff
@@ -1100,11 +1258,17 @@ class MoveAssignmentOperator extends Operator {
...
+ * template<class T>
+ * void myTemplateFunction(T t) {
+ *   ...
+ * }
+ * ```
  *
  * This includes function declarations which are immediately preceded by
- * `template <...>`, where the "..." part is not empty, and therefore it
- * does not include:
+ * `template <...>`, where the "..." part is not empty, but does not include:
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/Function.qll]: Why this change? Do you think one of the bullets below is not a logical consequence of this sentence?

If this sentence is indeed a precise description of `TemplateFunction`, then we can change "includes" to "comprises" to clarify that there are no other `TemplateFunction`s than those described here.

File: cpp/ql/src/semmle/code/cpp/Function.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/Function.qll]: I just had a second look at this, and I can see that it's technically legal to have `const` there, but I'd still argue that we should remove the `const`. It makes little sense to declare an rvalue reference as const since the whole purpose is to move out of it, which will lead to modifying it in non-trivial cases.

File: cpp/ql/test/library-tests/functions/functions/functions.cpp [hunk 4]
```diff
@@ -30,4 +30,12 @@ class Table {
...
 };
 
-
+class MyClass
+{
+public:
+	MyClass();
+	MyClass(int from);
+	MyClass(const MyClass &from);
+	MyClass(const MyClass &&from);
```
Comment (jbj) [inline cpp/ql/test/library-tests/functions/functions/functions.cpp]: Please either remove `const` or add another move constructor that doesn't have `const`.

File: cpp/ql/src/semmle/code/cpp/Function.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/Function.qll]: I haven't really played with move constructors or thought about them much, just copied the syntax I'd seen elsewhere (which probably did have a non-const version alongside).  I've removed `const`.

File: cpp/ql/src/semmle/code/cpp/Function.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/Function.qll]: Done.

File: cpp/ql/src/semmle/code/cpp/Function.qll (hunk 3 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/Function.qll]: Yes, I had a problem with "therefore" because you can't deduce that things aren't included from a statement that appears to be describing only some of the things that are.

I think it's supposed to describe all of them.  Changed to "comprises" and back to "therefore".

File: cpp/ql/test/library-tests/functions/functions/functions.cpp (hunk 4 above)
Comment (geoffw0) [reply cpp/ql/test/library-tests/functions/functions/functions.cpp]: Done.
